engine/table/src/main/java/io/deephaven/engine/table/impl/indexer/RowSetIndexer.java

```diff
@@ -243,24 +251,25 @@ private static Map<Object, RowSet> lookupMapping(
     private static void collectGrouping(
             final RowSet rowSetRowSet,
             final UnaryOperator<RowSet> rowSetOp,
-            final WeakHashMap<List<ColumnSource>, MappingInfo> mappings,
-            final WeakHashMap<List<ColumnSource>, MappingInfo> ephemeralMappings,
+            final WeakHashMap<List<ColumnSource<?>>, MappingInfo> mappings,
+            final WeakHashMap<List<ColumnSource<?>>, MappingInfo> ephemeralMappings,
             final BiConsumer<Object, RowSet> resultCollector,
             final TupleSource tupleSource,
-            final List<ColumnSource> keyColumns) {
+            final List<ColumnSource<?>> keyColumns) {
         if (keyColumns.isEmpty()) {
             resultCollector.accept(EmptyTuple.INSTANCE, rowSetRowSet.copy());
-        } else if (keyColumns.size() == 1 && keyColumns.get(0).getGroupToRange() != null) {
+        } else if (keyColumns.size() == 1 && keyColumns.get(0).hasGrouping()) {
             @SuppressWarnings("unchecked")
-            final Map<Object, RowSet> sourceGrouping = keyColumns.get(0).getGroupToRange();
-            for (final Map.Entry<Object, RowSet> objectRowSetEntry : sourceGrouping.entrySet()) {
-                final RowSet resultRowSet = rowSetOp.apply(objectRowSetEntry.getValue());
-                if (resultRowSet.size() > 0) {
-                    resultCollector.accept(objectRowSetEntry.getKey(), resultRowSet);
+            final GroupingBuilder builder = keyColumns.get(0).getGroupingBuilder();
+            final Map<Object, RowSet> groupings = builder.buildGroupingMap();
+            groupings.forEach((key, value) -> {
+                final RowSet resultIndex = rowSetOp.apply(value);
```

Comment:
I'm suspicious that rowSetOp is always a no-op, or intersecting (which could be done with strict building), and that resultCollector is being used to make a map or a table, potentially unnecessarily.